### ISSUE TITLE
AutoKeras 1.0 much slower than 0.4 on Google Colab

### ISSUE THREAD

**Issue by pezdorado** [keras-team/autokeras#834]
When I try to run a simple MNIST example on Google Colab with GPU with Autokeras 0.4 it runs very fast (1 epoch of the first model takes < 2 s) but with 1.0 it runs much slower (1 epoch of the first model takes > 80 s). When I disable the GPU 0.4 runs as slow as 1.0 which suggests that 1.0 isn’t using the GPU. How can I make Autokeras 1.0 run as fast as 0.4 with GPU?

To reproduce go to [colab.research.google.com](url), choose a Python 3 runtime with GPU accelerator, and execute the following

0.4 code

```
%tensorflow_version 1.x

!pip install autokeras

import autokeras
import tensorflow

( ( x, y ), validation_data ) = tensorflow.keras.datasets.mnist.load_data( )
model = autokeras.ImageClassifier( verbose = True )
model.fit( x, y )
```

1.0 code

```
%tensorflow_version 2.x

!pip install git+git://github.com/keras-team/keras-tuner@master#egg=keras-tuner
!pip install git+git://github.com/keras-team/autokeras@master#egg=autokeras

import tensorflow
import autokeras

( ( x, y ), validation_data ) = tensorflow.keras.datasets.mnist.load_data( )
model = autokeras.ImageClassifier( )
model.fit( x, y, validation_data = validation_data )
```

**Comment by TreeKat71** [comment]
In the 1.0, you have validation_data, so it does make sense

**Comment by pezdorado** [comment]
> In your 1.0 code, you have validation_data, so it does make sense.
> You still need to validate it, it takes more time.

The validation_data is only a small fraction of the training data ( 10000 versus 60000 samples I believe ) and does not explain the speed factor of > 40. When I use validation_split = 0.2 instead of validation_data the speed difference remains ( I understand 0.4 also implicitly uses a validation split but I cannot find what fraction ).

**Comment by jangminhyeok** [comment]
Same here.
I tried to run @pezdorado 's code with cpu and gpu, and runtime was almost same.
Also, I think trying to run it with gpu has some issues.
It just stopped at "4/Unknown - 3s 643ms/step - loss: 2.8299 - accuracy: 0.0625".
Btw It's ok to me, I will just use it with cpu.


2019-11-14 23:10:53.210017: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cudart64_100.dll
2019-11-14 23:10:56.531981: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library nvcuda.dll
2019-11-14 23:10:56.818911: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1618] Found device 0 with properties: 
name: GeForce GTX 1060 major: 6 minor: 1 memoryClockRate(GHz): 1.733
pciBusID: 0000:01:00.0
2019-11-14 23:10:56.819257: I tensorflow/stream_executor/platform/default/dlopen_checker_stub.cc:25] GPU libraries are statically linked, skip dlopen check.
2019-11-14 23:10:56.819826: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1746] Adding visible gpu devices: 0
2019-11-14 23:10:56.820227: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2
2019-11-14 23:10:56.823384: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1618] Found device 0 with properties: 
name: GeForce GTX 1060 major: 6 minor: 1 memoryClockRate(GHz): 1.733
pciBusID: 0000:01:00.0
2019-11-14 23:10:56.823687: I tensorflow/stream_executor/platform/default/dlopen_checker_stub.cc:25] GPU libraries are statically linked, skip dlopen check.
2019-11-14 23:10:56.824222: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1746] Adding visible gpu devices: 0
2019-11-14 23:10:57.511984: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1159] Device interconnect StreamExecutor with strength 1 edge matrix:
2019-11-14 23:10:57.512230: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1165]      0 
2019-11-14 23:10:57.512375: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1178] 0:   N 
2019-11-14 23:10:57.513114: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1304] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 4708 MB memory) -> physical GPU (device: 0, name: GeForce GTX 1060, pci bus id: 0000:01:00.0, compute capability: 6.1)
Epoch 1/1000
2019-11-14 23:11:17.213769: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cublas64_100.dll
2019-11-14 23:11:17.463753: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cudnn64_7.dll
2019-11-14 23:11:18.442970: W tensorflow/stream_executor/cuda/redzone_allocator.cc:312] Internal: Invoking ptxas not supported on Windows
Relying on driver to perform ptx compilation. This message will be only logged once.

      1/Unknown - 2s 2s/step - loss: 3.0134 - accuracy: 0.0625
      2/Unknown - 3s 1s/step - loss: 2.8319 - accuracy: 0.0469
      3/Unknown - 3s 846ms/step - loss: 2.8356 - accuracy: 0.0625
      4/Unknown - 3s 643ms/step - loss: 2.8299 - accuracy: 0.0625

**Comment by jhfjhfj1** [comment]
Thank you for the issue. We are trying to solve it.

**Comment by pezdorado** [comment]
Thank you for reopening this and trying to solve this.

In #847 it was suggested that this problem has to do with data augmentation.

I'm having a hard time understanding how this could explain the fact that without GPU 0.4 and 1.0 are equally fast but with GPU 0.4 is much faster and 1.0 is still as slow as without GPU.

Also I didn't know AutoKeras did image augmentation. I tried to find more information about this, and I found that the 0.4 ImageClassifier( ) has an augment parameter that defaults to False but the 1.0 ImageClassifier( ) doesn't seem to have an augment parameter. Does 1.0 do augmentation and if so is there a way to switch it off?

**Comment by jhfjhfj1** [comment]
@pezdorado You can turn off the augmentation by using the IO API of 1.0. The image block has that arg. Please refer to the tutorial on the official website. Thanks

**Comment by pezdorado** [comment]
I've tried the functional API with an ImageBlock( augment = False ) but it's as slow as the task API. I hope I applied the functional API correctly since the tutorial is quite sparse

> !pip install git+git://github.com/keras-team/autokeras@master#egg=autokeras
> 
> import tensorflow
> import autokeras
> 
> ( ( x, y ), validation_data ) = tensorflow.keras.datasets.mnist.load_data( )
> 
> inputs = autokeras.ImageInput( )
> blocks = autokeras.hypermodel.hyperblock.ImageBlock( augment = False )( inputs )
> outputs = autokeras.ClassificationHead( )( blocks )
> 
> model = autokeras.GraphAutoModel(
>     inputs = inputs,
>     outputs = outputs
> )
> 
> model.fit( x, y, validation_split = 0.1, batch_size = 128 )

Did I apply the functional API correctly and did I correctly turn off the augmentation? If so, then the problem does not seem to be caused by augmentation.

**Comment by jhfjhfj1** [comment]
@pezdorado You did it correctly. Thank you!

You can further break down the ImageBlock into [Normalization] + [ImageAugmentation] + [ConvBlock, ResNetBlock, XceptionBlock]. In this way, we can see which one is the slow one. 

I remember I tried just ConvBlock + ClassificationHead. It was fast.

**Comment by pezdorado** [comment]
@jhfjhfj1 Thank you for your answers. I measured the execution times of the different blocks for 1 epoch both with and without GPU

```
inputs = autokeras.ImageInput( )
# t1 = autokeras.hypermodel.preprocessor.Normalization( )( inputs )
# t1 = autokeras.hypermodel.preprocessor.ImageAugmentation( )( inputs )
# t1 = autokeras.hypermodel.block.ConvBlock( )( inputs )
# t1 = autokeras.hypermodel.block.XceptionBlock( )( inputs )
t1 = autokeras.hypermodel.block.ResNetBlock( )( inputs )
outputs = autokeras.ClassificationHead( )( t1 )

model = autokeras.GraphAutoModel(
    inputs = inputs,
    outputs = outputs
)
```

With GPU
Normalization( ): 85 s
ImageAugmentation( ): 278 s
ConvBlock( ): 11 s
ResNetBlock( ): 45 s
XceptionBlock( ): 42 s

Without GPU:
Normalization( ): 33 s
ImageAugmentation( ): 80 s
ConvBlock( ): 111 s
ResNetBlock( ): 3882 s
XceptionBlock( ): 2659 s

So on 1.0 the GPU makes ConvBlock( ), ResNetBlock( ) and XceptionBlock( ) much faster but Normalization( ) and ImageAugmentation( ) slower. However the 1.0 ConvBlock( ) is still is much slower than the 0.4 full ImageClassifier( ) that only takes 2 s for 1 epoch of the initial model with GPU.

**Comment by KalidindiMounika** [comment]
There is no option to export/save a model in autokeras 1.0. How can we save the model for deployment ?

**Comment by haifeng-jin** [comment]
@KalidindiMounika We have the function `export` in the new release now.
It can export everything except for the preprocessors.

**Comment by haifeng-jin** [comment]
This issue has been breakdown to multiple parts in the AutoKeras Project.
After adopting the preprocessing layers the issue should be resolved.
